File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
from itertools import combinations
from collections import defaultdict

def solution(orders, course):
    
    d = [defaultdict(int) for _ in range(course[-1] + 1)]
    
    for order in orders:
        for c in course:
            if c <= len(order):
                for a in combinations(sorted(list(order)), c):
                    d[c][a] += 1
    
    answer =[]
    for c in course:
        if d[c]:
            max_val = max(d[c].values())
            if max_val >= 2:
                answer += ["".join(k) for k in d[c].keys() if d[c][k] == max_val]
    
    return sorted(answer)
        
        


        
        

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public String[] solution(String[] orders, int[] course) {
        List<String> answer = new ArrayList<>();
        
        
        for (int i = 0; i < course.length; i++) {
            Map<String, Integer> o = new HashMap<>();
            for(String order : orders) {
                char[] a = order.toCharArray();
                Arrays.sort(a);
                order = new String(a);
                List<String> combs = find(order, course[i]);
                for (String comb : combs) {
                    o.merge(comb, 1, (old, value) -> old + 1);
                }
            }
            if (o.isEmpty()) {
                continue;
            }
            
            int m = o.entrySet().stream()
                .max(Map.Entry.comparingByValue())
                .get()
                .getValue();
            
            if (m < 2) {
                continue;
            }
            
            for (Map.Entry<String, Integer> entry : o.entrySet()) {
                if (entry.getValue() == m) {
                    answer.add(entry.getKey());
                }
            }
            
        }
        
        answer.sort(Comparator.naturalOrder());
        
        
        return answer.toArray(new String[0]);
    }
    
    public List<String> find(String origin, int course) {
        if (origin.length() < course) {
            return List.of();
        }
        return findCourse(origin, "", 0, course);
    }
    
    public List<String> findCourse(String origin, String current, int index, int course) {        
        if (course == 0) {
            return List.of(current);
        }
        
        List<String> result = new ArrayList<>();
        
        for (int i = index; i < origin.length() - course + 1; i++) {
            result.addAll(findCourse(origin, current + origin.charAt(i), i + 1, course - 1));
        }
        
        return result;
    }
}